src/funcx_common/messagepack/message_types/ep_status_report.py

```diff
@@ -1,77 +1,21 @@
+
 import typing as t
 import uuid
+from dataclasses import dataclass
 
-from ..common import Message, MessageType
-from ..exceptions import InvalidMessageError, InvalidMessagePayloadError
+from ..common import Message
 
 
+@dataclass
 class EPStatusReport(Message):
     """
     Status report for an endpoint, sent from Endpoint to Forwarder.
 
     Includes EP-wide info such as utilization, as well as per-task status information.
     """
 
-    message_type = MessageType.EP_STATUS_REPORT
-
-    def __init__(
-        self,
-        endpoint_id: str,
-        ep_status_report: t.Dict[str, t.Any],
-        task_statuses: t.Dict[str, t.Any],
-    ) -> None:
-        if not (isinstance(ep_status_report, dict) and isinstance(task_statuses, dict)):
-            raise InvalidMessageError(
-                "EPStatusReport inner json data was improperly shaped"
-            )
-
-        self.endpoint_id = endpoint_id
-        self.ep_status = ep_status_report
-        self.task_statuses = task_statuses
-
-    @property
-    def endpoint_id_bytes(self) -> bytes:
-        return uuid.UUID(self.endpoint_id).bytes
-
-    def get_v0_body(self) -> bytes:
-        # NOTE: it's important that this is done as a dumps(...).encode("ascii")
-        # rather than the more normal "dumps(..., ensure_ascii=True)"
-        # the reason is that this is the method of encoding used in `funcx-endpoint` and
-        # we must be sure that
-        # 1. messages written by this module can be read there
-        # 2. messages written by that module can be read here
-        #
-        # (1) is ensured by this method of encoding in this function
-        # (2) is ensured by testing that `load_v0_body` is an inverse for this method
-        jsonified = json.dumps(
-            [self.ep_status, self.task_statuses],
-            separators=(",", ":"),
-            sort_keys=True,
-        ).encode("ascii")
-        return self.endpoint_id_bytes + jsonified
-
-    @classmethod
-    def load_v0_body(cls, buf: bytes) -> "EPStatusReport":
-        if len(buf) <= 16:
-            raise InvalidMessagePayloadError("EPStatusReport body was too short")
-
-        # this cannot fail because any 16-byte sequence can be converted to a UUID
-        endpoint_id = str(uuid.UUID(bytes=buf[:16]))
-
-        buf = buf[16:]
-        jsonified = buf.decode("ascii")
-        try:
-            loaded_json = json.loads(jsonified)
-        except ValueError as e:
-            raise InvalidMessagePayloadError(
-                "EPStatusReport body did not contain JSON section"
-            ) from e
-
-        if not (isinstance(loaded_json, list) and len(loaded_json) == 2):
-            raise InvalidMessagePayloadError(
-                "EPStatusReport json data was improperly shaped"
-            )
-
-        ep_status, task_statuses = loaded_json
-        return cls(endpoint_id, ep_status, task_statuses)
+    message_type = "ep_status_report"
+    endpoint_id: str | uuid.UUID
```

Comment:
It's not clear to me why this is a union of two different types - is an endpoint ID an opaque string token or is it a subset of that that is UUIDs? Do I have to expect both things to be true whereever I use an endpoint_id in funcx?

Reply:
or is this some hack annotation trying to trick the validator code into running the "subset-of-strings-that-is-also-uuid" validator code by making weird type annotations?

Reply:
The intention was to support strings which are UUIDs, not other strings.
I expect we have a lot of cases like in which we have a string which we know must represent a UUID. I wanted to support those cases without forcing callers to do the conversion themselves.

Do you think it's better to take the UUID type and insist that anything being passed through this layer be a UUID? It would make any validation look more like a type guard.

Reply:
I'm not sure if fixing any problems here are in scope for this PR.

 I would expect if I'm being told I'm getting a dict of UUIDs that I wouldn't need to concern myself with things like "oh the wire protocol capitalised things differently to the way my user pasted a UUID into their source file", and that pushes me to want `UUID` objects everywhere.

I'm also fine with IDs being treated as not UUIDs, but opaque string tokens - those have different `==` behaviour, so affect use as dictionary keys. But in that case, we should not be claiming they are UUIDs (even if we happen to use a UUID generator as part of generating the opaque string).

Reply:
Now that I've done the rewrite to use pydantic, this whole issue is "taken care of". The pydantic behavior is that if you declare a field as type `uuid.UUID`, it will accept str, bytes, int, and UUID values, and attempt to coerce them all to UUID.

I'm not totally happy with all of the pydantic behaviors, but the most important thing is satisfied: if you have an EpStatusReport message object, then `message.endpoint_id` will be of type `UUID`.

I'm going to mark this as resolved, but please feel free to un-resolve if we should discuss further.